### ISSUE TITLE
Grow beyond Docker Swarm

### ISSUE THREAD

**Issue by kallisti5** [haiku/infrastructure#66]
When we rolled our containerized infrastructure out in 2017, we chose Docker Swarm + Docker Compose due to its simplicity.

However, since [then Kubernetes has grown by leaps and bounds](https://thenewstack.io/data-says-kubernetes-deployment-patterns/) to become the "standard" for > 50% of cloud organizations.   We use Rexray in our Docker swarm to have "Kubernetes like managed persistent volumes"... however [Rexray](https://github.com/rexray/rexray) is essentially a dead project creating risk.

The *BIG* downside is cloud costs have been steadily increasing everywhere.  A reasonable sized 3-node Kubernetes cluster for our workloads at Digital Ocean is around $144-$240 / month.   K8S + DO Managed load balancers are likely extra as well. (today we spend ~$124.03 / month on one single big node and a small IPFS gateway node in Germany)

**Pros:**
  * More standardized stack
  * Limitless scaling.  We could fire up a few additional K8S nodes before big releases like R1
  * Worker nodes can be rebooted with limited impact to running services
  * We gain 100% uptime upgrades (badly needed for things like software repositories... though it may not matter as much with IPFS?)

**Cons:**
  * Cost (more smaller nodes over one larger node)

Other solutions like Hashicorp Nomad have been suggested... but I don't know these solutions well enough to make an informed decision.   AWS + GCP + Digital Ocean + Vultr all support managed K8S, and none of them support managed Nomad.

The cheaper solution would be going back to dedicated server instances or colocation running Kubernetes... but these all require manpower to manage (which we're short on).

tldr;  Does the increase in price justify the improved ease of management?

**Comment by kallisti5** [comment]
For the "lets keep what we have right now" argument:

Our current infrastructure [metadata](https://github.com/haiku/infrastructure/blob/master/ci.yaml#L21) needs rewritten for Traefik v2 (as V1 is seeing fewer and fewer security updates) *OR* we need to design an alternative reverse proxy (ingress controller in K8S speak) which works with Docker Swarm and has auto letsencrypt support.

My thought is we need to do a major refactoring either way... so lets do it right.

**Comment by kallisti5** [comment]
As seen in https://github.com/haiku/infrastructure/pull/67, 

[Docker Swarm and Docker compose don't internally support IPv6](https://docs.docker.com/compose/compose-file/compose-file-v3/#ipv4_address-ipv6_address).  This is a huge issue since we can't filter inbound IPv6 requests by IP (everything gets squished into a private IPv4 NAT address)

**Comment by nielx** [comment]
Let me caveat by saying that I am not qualified to make the call. But I do think it should be made with all the information, and with clear goals in mind.

We previously switched from Docker compose to Docker Swarm because of a theoretical future where we have to increase the number of nodes that serve a service. I am sceptical that this really is a must have and not a nice to have under the current conditions. So far our services seem to be able to handle what is coming to them. It is also not the case that our infrastructure powers critical web services. It would be a major inconvenience if our services were down, but nobody dies and no SLAs are breached. 

Attached to that, we need to first see who we have in the infrastructure team, and perhaps consider finding a junior to support us. 

Finally, I feel in my water that your mind is already set on it and that you are eager to make the switch as soon as possible. Pause. Make some architecture diagrams. Gather some learning materials for us noobs. And then let's discuss a plan.

**Comment by kallisti5** [comment]
>  But I do think it should be made with all the information, and with clear goals in mind.

Agree

> We previously switched from Docker compose to Docker Swarm because of a theoretical future where we have to increase the number of nodes that serve a service. 

Technically we switched from Docker to Docker Swarm in ~2018.  Compose is the language to configure docker or swarm.

>  I am sceptical that this really is a must have and not a nice to have under the current conditions. So far our services seem to be able to handle what is coming to them. It is also not the case that our infrastructure powers critical web services. It would be a major inconvenience if our services were down, but nobody dies and no SLAs are breached.

That's 100% fair.  To be honest, my biggest concern pushing this is the package repositories and HaikuDepot.   Pretty much every installation of Haiku interfaces with these services. Today we don't have any redundancy around these services and their availability. If we do maintenance to them or the ingress, they go down briefly

As we grow we're going to need to support a slowly growing traffic load, as well as a slowly growing storage consumption.

Some numbers today:
  * ~3-7 TiB total a month of data transfer
  * ~230 GiB on limerick of storage, ~1 TiB of storage at wasabi
All of these numbers are 1-2x greater than when the original infrastructure was deployed.

Considerations:
  * Docker Swarm doesn't support IPv6 internally
  * We use Rexray for dynamic volute attachment in Swarm. It's now abandoned.
  * Traefik 1.x is no longer supported.
    * We have to rewrite pretty much all of our docker compose files, lets take the opportunity to improve things as well.
  * Wasabi has already flagged and disabled our services due to traffic usage once. They may do it again.
    * Wasabi hosts nightly images, haiku's package repositories, and some inc. stuff.
  * I deployed a small IPFS gateway node in Germany, this was to help distribute load, but also add more bandwidth to our DigitalOcean account's "pool of bandwidth".


As for training. kubernetes works pretty much identically to docker in swarm mode:

* **Enter a container:**
  * **Docker / Docker Swarm:**  ```docker exec -it (hash) /bin/bash -l```
  * **Kubernetes:** ```kubectl exec -it (pod name) -- /bin/bash-l```
* **Updating container specifications**
  * **Docker Swarm:** ```DOMAIN=haiku-os.org docker stack deploy -c thing.yaml thing```
  * **Kubernetes:** ```kubectl apply -f thing.yaml```
* **Updating containers**
  * **Docker Swarm:** ```docker service update --image docker.io/haikuporter/buildmaster:1.2.5 --force ci_buildmaster_master_x86_64```
  * **Kubernetes:** ```kubectl set image deployment/ci_buildmaster www=docker.io/haikuporter/buildmaster:1.2.5```
*  **Scaling servicing pods**
  * **Docker Swarm:** ```docker service scale ci_buildmaster=3```
  * **Kubernetes:** ```kubectl scale --replicas=3 deployments/ci_buildmaster```

Here's what we get with a potential Kubernetes Change:
* We don't pay for master nodes
* Digital Ocean manages the Kubernetes clusters for us
  * If we need to add more nodes, we tell Digital Ocean, and our cluster scales out horizontally
* Digital Ocean manages persistent data volumes and their attachments to VM's
  * If we need more space, we adjust the persistent volume size and kubernetes talks to the Digital Ocean API to add more storage.

There's a pretty good high level rundown of a hosted kubernetes cluster here:
https://www.youtube.com/watch?v=7bA0gTroJjw

**Comment by nielx** [comment]
Thank you for the long and comprehensive reply!

> > We previously switched from Docker compose to Docker Swarm because of a theoretical future where we have to increase the number of nodes that serve a service.
> 
> Technically we switched from Docker to Docker Swarm in ~2018. Compose is the language to configure docker or swarm.

See, I need training :-P

> As we grow we're going to need to support a slowly growing traffic load, as well as a slowly growing storage consumption.
> 
> Some numbers today:
> 
> * ~3-7 TiB total a month of data transfer
> * ~230 GiB on limerick of storage, ~1 TiB of storage at wasabi
>   All of these numbers are 1-2x greater than when the original infrastructure was deployed.

Right, from this perspective it does seem like we are going to hit a wall sooner or later. The biggest driver of that traffic and data storage are the images and HaikuPorts repositories.  Would it make sense to investigate specialized options for these repository-based tools rather than blanket migrate everything to a new host? Like I said before, I am not "in the market", but I imagine content distribution has different tools available (with different pricing) than the more processing-dependent tools like Trac and Git.

> Considerations:
> 
> * Docker Swarm doesn't support IPv6 internally
> * We use Rexray for dynamic volute attachment in Swarm. It's now abandoned.
> * Traefik 1.x is no longer supported.
>   
>   * We have to rewrite pretty much all of our docker compose files, lets take the opportunity to improve things as well.
> * Wasabi has already flagged and disabled our services due to traffic usage once. They may do it again.
>   * Wasabi hosts nightly images, haiku's package repositories, and some inc. stuff.

Points taken. Doing nothing is not an option.

> * I deployed a small IPFS gateway node in Germany, this was to help distribute load, but also add more bandwidth to our DigitalOcean account's "pool of bandwidth".

IPFS is too much in its infancy to rely on at this moment, so IMO it would not be wise to assume it will be there to help out. Even if IPFS is part of the future, it is not unlikely that we will still need to host the most up to date IPFS master node, which might be a naturally high-load node for that reason. So the best IPFS offers right now, is maybe give us the requirement to be able to scale down if the demand for data traffic reduces in the future.

> There's a pretty good high level rundown of a hosted kubernetes cluster here:
> https://www.youtube.com/watch?v=7bA0gTroJjw
> 
> > Finally, I feel in my water that your mind is already set on it and that you are eager to make the switch as soon as possible. Pause. Make some architecture diagrams. Gather some learning materials for us noobs. And then let's discuss a plan.
> 
> My mind is... but i'm not the only sysadmin. That's why I opened this ticket to discuss :-)

Good to hear! Well I will be with you on the trip. Let me do some learning. I can help out with making the plan and present it to the board.

**Comment by kallisti5** [comment]
> IPFS is too much in its infancy to rely on at this moment, so IMO it would not be wise to assume it will be there to help out. Even if IPFS is part of the future, it is not unlikely that we will still need to host the most up to date IPFS master node, which might be a naturally high-load node for that reason. So the best IPFS offers right now, is maybe give us the requirement to be able to scale down if the demand for data traffic reduces in the future.

I actually agree.   It has been working as a backup (for example, it's my "wasabi blocks us again" last-resort fallback plan) and offers some compelling stuff (you haven't lived until you have done pkgman update from a local IPFS node with all the data on your network), but i've encountered *way* too many issues around data availability for us to depend on it as our primary CDN.

It was an experiment, but it didn't scale to our primary distribution mechanism.  I'm going to keep the experiment going as it really only depends on a Raspberry Pi 4 at my house, but not really push for it to be our primary content delivery mechanism.... it just doesn't support big data sets well.

> Right, from this perspective it does seem like we are going to hit a wall sooner or later. 

Yup.   I've been trying to balance cost with ease of management for a while.

* Colocation will give us the most value for our money.
  * However it requires a full-time dedicated admin.
  * Our history shows we really can't support this model.  I'd like to not sysadmin all of the time for Haiku as a full time job. If I get hit by a bus, Haiku should have time for others to train.

* Cloud costs more, but it removes a hours required for day-to-day management.
  * I'll give an example.  Trac likes to crap out and die (I think someone is trying to abuse it from the logs)
  * Docker swarm has been transparently restarting Trac every few days when it goes down.

Ideally, we should really only need to mess with stuff when we want to make changes... that's my end goal :-).   Historically the sysadmin position has been reacting to outages, space issues, patching, etc.

There's a lot of "woowoo" behind cloud / devops stuff. But there are benefits if you make well thought out systems and not just try to buy products to stuff into your stack.

**Comment by kallisti5** [comment]
> Right, from this perspective it does seem like we are going to hit a wall sooner or later. The biggest driver of that traffic and data storage are the images and HaikuPorts repositories. Would it make sense to investigate specialized options for these repository-based tools rather than blanket migrate everything to a new host? Like I said before, I am not "in the market", but I imagine content distribution has different tools available (with different pricing) than the more processing-dependent tools like Trac and Git.

Realized I never replied to this :-)

I kind of already made a custom solution to cover the repos.  Haiku's packages are uploaded to s3 storage, and [hpkgbouncer](https://github.com/haiku/hpkgbouncer) redirects people from eu.hpkg.haiku-os.org to wasabi via 302 redirects.

The idea was to offload all the "big stuff" to buckets.   While this worked (and was a reasonable price under wasabi), they're back-peddling their unlimited egress (outbound) bandwidth which makes me a bit nervous.  We could toss a reverse caching proxy or something in-front of our buckets :thinking:.

None of this works with haikuporter buildmaster packages today though.  We need to add s3 support to haikuporter buildmaster

**Comment by nielx** [comment]
I think in general it makes sense to offload some of the server maintenance and focus on providing our services using containers. That's really what I have been doing anyway, so it fits in my way of working. I also think that this is probably the 'less fun' part anyway, and I think there is a good case to be made to have us focus on providing functionality and let others think about maintaining the underlying system.

Should we set up a meeting with @waddlesplash and talk it through?

**Comment by kallisti5** [comment]
It's mostly you and me @nielx  doing the sysadmin stuff :-)
When @waddlesplash is an employee though, we can boss him around and make him do stuff though like learning k8s :thinking: 

Still trying to get his employment setup via the BoD. Need to reach quorum in a meeting which hasn't happened yet :-|

**Comment by kallisti5** [comment]
> I also think that this is probably the 'less fun' part anyway, and I think there is a good case to be made to have us focus on providing functionality and let others think about maintaining the underlying system.

Converting stuff to Kubernetes workloads is my day job.... it's actually pretty fun :-D

What's pretty cool about managed Kubernetes is there really won't be any servers to maintain anymore.  Pretty much everyone will just manage containers and infrastructure workloads via kubectl from their desktops, or use neat GUI tools like https://k8slens.dev which is awesome for quickly getting an idea on what's going on.

The stuff above about @waddlesplash is outdated now.  He's an employee! :fireworks:

**Comment by kallisti5** [comment]
So.. looking at Digital Ocean's managed kubernetes offerings...

![k8s](https://user-images.githubusercontent.com/100110/137743201-70a927b9-01b6-4948-b756-dbb54610f476.png)
![volume](https://user-images.githubusercontent.com/100110/137743218-694a9d6e-ebd7-4c76-9397-ac800e407872.png)
![load_balancer](https://user-images.githubusercontent.com/100110/137743509-8e0d3947-8ae8-4b89-9b5b-7dbd6d690b21.png)

That's roughly $210-$230 / month minimum to move to a kubernetes deployment with some redundancy. Today we pay $124.03. so ~$86 / month increase

So the question becomes.. does +$86   / month justify:
* Redundancy + zero downtime upgrades
* No more servers to manage
* Ability to scale horizontally

**Comment by kallisti5** [comment]
Hmmm.    Vultr is now offering NVMe block storage and managed kubernetes at a compelling price at their Amsterdam datacenter

![managed-1](https://user-images.githubusercontent.com/100110/151898976-0f1d0a1a-de30-47c7-a9e2-d5e25c35e4eb.png)
![storage-1](https://user-images.githubusercontent.com/100110/151898988-538b21e7-6298-47a0-b27a-fbdc03dd0196.png)

That's:

* $80 / mo for 2x k8s nodes, 8GiB of ram each - $120 / mo for 3x k8s nodes 8GiB of ram each
* ~ $40 / mo for 400GiB of NVMe storage
* $10 / mo for 1 load balancer 

So around $130 - $170 @ Vultr for a similar config :thinking:

**Comment by kallisti5** [comment]
done
